### PR TITLE
Admin panel

### LINKED ISSUE #3: Admin panel
Create admin panel.
Curator can edit schedule events, add locations, manage users.

### PR DESCRIPTION
**Pull request by sparksofl**
Close #3
Feature 3 Admin panel
Admin panel was created. Curator can manage users, events, locations.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -20,6 +20,4 @@
 .ruby-gemset
 
 config/database.yml
-app/views/layouts/.fuse_hidden00002b640000001b
-app/views/layouts/.fuse_hidden00002b640000001c
-app/views/layouts/.fuse_hidden00002b640000001d
+app/views/layouts/*fuse_hidden*
```
Comment (brytannia) [inline .gitignore]: Nice :+1:

File: app/admin/event.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+ActiveAdmin.register Event do
+	permit_params :datetime, :subject, :confirmed, :location_id
```
Comment (brytannia) [inline app/admin/event.rb]: don't forget about tabs:)

File: app/views/layouts/.fuse_hidden00002b640000001a [hunk 3]
```diff
@@ -0,0 +1,24 @@
+doctype html
```
Comment (brytannia) [inline app/views/layouts/.fuse_hidden00002b640000001a]: FYI, other hidden file

File: app/models/event.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+class Event < ActiveRecord::Base
+	has_many :users
```
Comment (brytannia) [inline app/models/event.rb]: as Event and User are connected by UserEvent model, you can't just connect these two model as you did. Check this article http://guides.rubyonrails.org/association_basics.html#the-has-many-through-association out for know about through association.

File: app/models/user_event.rb [hunk 5]
```diff
@@ -0,0 +1,4 @@
+class UserEvent < ActiveRecord::Base
+	belongs_to :users
```
Comment (brytannia) [inline app/models/user_event.rb]: user and event should be singular because 1 UserEvent can be connected with only 1 User and Event.

**Comment by brytannia** [comment]
@sparksofl if you finished task, please, assign me on PR and add label 'review' to this pr

**Comment by sparksofl** [comment]
@brytannia It's not done yet. I'll add 'review' after customizing panel

File: app/admin/dashboard.rb [hunk 6]
```diff
@@ -16,10 +14,20 @@
...
-        panel "Info" do
-          para "Welcome to ActiveAdmin."
+        panel 'Info' do
+          para 'Welcome to ActiveAdmin.'
         end
       end
     end
+
+    section 'This month' do
+      table_for Event.all.where('datetime > ? and datetime < ?',
```
Comment (brytannia) [inline app/admin/dashboard.rb]: .all method is redundant

File: db/schema.rb [hunk 7]
```diff
@@ -31,6 +31,28 @@
   add_index "active_admin_comments", ["namespace"], name: "index_active_admin_comments_on_namespace", using: :btree
   add_index "active_admin_comments", ["resource_type", "resource_id"], name: "index_active_admin_comments_on_resource_type_and_resource_id", using: :btree
 
+  create_table "events", force: :cascade do |t|
+    t.string   "subject"
```
Comment (brytannia) [inline db/schema.rb:35]: why Events don't have date and time?

File: db/migrate/20160208165815_create_events.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class CreateEvents < ActiveRecord::Migration
+  def change
+    create_table :events do |t|
+      t.timestamp :datetime
```
Comment (brytannia) [inline db/migrate/20160208165815_create_events.rb]: the type should be datetime

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -1,6 +1,6 @@
 class ApplicationController < ActionController::Base
   before_action :configure_permitted_parameters, if: :devise_controller?
-  
+  before_action :debug
```
Comment (brytannia) [inline app/controllers/application_controller.rb]: do not forget to remove :octocat:

File: app/models/event.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+class Event < ActiveRecord::Base
+  has_many :user_events
+  has_many :users, through: :user_events
+  belongs_to :location
+
+  accepts_nested_attributes_for :users, allow_destroy: true
```
Comment (brytannia) [inline app/models/event.rb]: from my point of view, allow_destroy shouldn't be there. What do you think?

File: app/assets/javascripts/application.js [hunk 11]
```diff
@@ -13,4 +13,4 @@
 //= require jquery
 //= require jquery_ujs
 //= require turbolinks
-//= require_tree .
+//= require_tree .
```
Comment (brytannia) [inline app/assets/javascripts/application.js]: endline

File: app/assets/stylesheets/active_admin.scss [hunk 12]
```diff
@@ -14,4 +14,4 @@
 // Overriding any non-variable SASS must be done after the fact.
 // For example, to change the default status-tag color:
 //
-//   .status_tag { background: #6090DB; }
+//   .status_tag { background: #6090DB; }
```
Comment (brytannia) [inline app/assets/stylesheets/active_admin.scss]: endline

File: app/admin/user.rb [hunk 13]
```diff
@@ -1,17 +1,28 @@
...
+    column :email
+    actions
+  end
+  form do |f|
+    f.inputs 'User' do
+      f.input :first_name
+      f.input :last_name
+      f.input :speaker
+      f.input :email
+      f.input :password
```
Comment (brytannia) [inline app/admin/user.rb:19]: let's leave this as it is. We'll remove password from this form later

File: app/admin/event.rb [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  scope :today do |events|
+    events.where('? < datetime and datetime < ?',
+                 DateTime.now.to_date - 1.days, DateTime.now.to_date + 1.days)
+  end
+  scope :this_week do |events|
+    events.where('datetime > ? and datetime < ?',
+                 DateTime.now.to_date, 1.week.from_now)
+  end
+  scope :past do |events|
+    events.where('datetime < ?', DateTime.now.to_date)
```
Comment (brytannia) [inline app/admin/event.rb]: Read about DateTime methods that provide enhanced support. http://api.rubyonrails.org/classes/DateTime.html

for this scope you can use DateTime.now.beginning_of_day

File: app/admin/event.rb [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    end
+    f.actions
+  end
+
+  filter :confirmed, as: :check_boxes
+
+  scope :all, default: true
+  scope :today do |events|
+    events.where('? < datetime and datetime < ?',
+                 DateTime.now.to_date - 1.days, DateTime.now.to_date + 1.days)
```
Comment (brytannia) [inline app/admin/event.rb]: few comments. also read comment below.
1. for such queries you can use Range:
events.where(datetime: one_date..second_date)
2. better to use other scope for querying today events:
DateTime.now.beginning_of_day, DateTime.now.end_of_day

File: db/migrate/20160208165815_create_events.rb (hunk 8 above)
Comment (sparksofl) [inline db/migrate/20160208165815_create_events.rb]: But Postgresql doesn't have any datetime as a datatype
http://www.postgresql.org/docs/9.1/static/datatype-datetime.html

Comment (brytannia) [inline db/migrate/20160208165815_create_events.rb]: but we use ActiveRecord wrapper which has datetime type.

File: app/models/event.rb (hunk 10 above)
Comment (sparksofl) [inline app/models/event.rb]: As I understand it's for nested form: when admin adds users to the speakers -- he also can destroy one of them (remove from the speakers of a certain event). Not sure. It's uncompleted part.

Comment (brytannia) [inline app/models/event.rb]: allow_destroy option allow to delete object from db. that's definitely not what we want to do.

File: db/migrate/20160208165815_create_events.rb (hunk 8 above)
Comment (sparksofl) [inline db/migrate/20160208165815_create_events.rb]: Should I create new migrations to delete that column and create new with the right type?

Comment (brytannia) [inline db/migrate/20160208165815_create_events.rb]: change this or create new one - it's up to you.

File: app/admin/event.rb [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  permit_params :subject, :datetime, :confirmed, :location_id,
+                :user_events, :users
+  before_filter :set_users, only: [:update]
+
+  controller do
+    def set_users
+      @event = Event.find(params[:id])
+      @event.users = []
+      params[:event][:user_ids].each do |id|
+        @event.users << User.find(id) if id != ''
```
Comment (brytannia) [inline app/admin/event.rb]: in case of if and not statement, use unless

File: app/admin/dashboard.rb [hunk 17]
```diff
@@ -16,10 +14,20 @@
...
-        panel "Info" do
-          para "Welcome to ActiveAdmin."
+        panel 'Info' do
+          para 'Welcome to ActiveAdmin.'
         end
       end
     end
+
+    section 'This month' do
+      table_for Event.where('datetime > ? and datetime < ?',
```
Comment (brytannia) [inline app/admin/dashboard.rb]: why don't you use Range?

File: app/admin/event.rb [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+      f.semantic_errors(*f.object.errors.keys)
+      f.input :subject
+      f.input :datetime, as: :datetime_picker
+      f.input :location
+      f.input :confirmed
+
+      f.input :users,
+              collection: User.all.map { |user| [user.email, user.id] },
+              as: :select, multiple: true
+      ##########
```
Comment (brytannia) [inline app/admin/event.rb]: please, remove this commented code

File: app/models/user.rb [hunk 19]
```diff
@@ -1,8 +1,13 @@
...
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
-         
-  def is_admin?
-    self.admin
+
+  validates_presence_of :first_name, :last_name, :email,
+                        :password, :password_confirmation
+
+  def admin?
```
Comment (brytannia) [inline app/models/user.rb]: this method is predefined by rails itself http://makandracards.com/makandra/607-automagical-methods-for-boolean-attributes-in-activerecord (so you can delete it)

File: app/admin/dashboard.rb (hunk 17 above)
Comment (sparksofl) [inline app/admin/dashboard.rb]:
```
events.where(datetime: DateTime.now.beginning_of_day...
                 DateTime.now.beginning_of_day + 1.days)
```

smth like that?

Comment (brytannia) [inline app/admin/dashboard.rb]: events.where(datetime: DateTime.now.beginning_of_day..
                 DateTime.now.end_of_day)
note: in this case we use including range

Comment (brytannia) [inline app/admin/dashboard.rb]: DateTime.now.beginning_of_day + 1.days -> for future this exp can be replaced with 1.day.from_now

**Comment by brytannia** [comment]
Nice job! as usual :wink:

**Pull request merged**